crates/bevy_reflect/src/serde/ser.rs

```diff
@@ -177,6 +196,14 @@ pub struct TypedReflectSerializer<'a> {
 
 impl<'a> TypedReflectSerializer<'a> {
     pub fn new(value: &'a dyn Reflect, registry: &'a TypeRegistry) -> Self {
+        #[cfg(feature = "debug_stack")]
+        TYPE_INFO_STACK.set(crate::type_info_stack::TypeInfoStack::new());
+
+        TypedReflectSerializer { value, registry }
+    }
```

Comment:
Since `TypedReflectSerializer` is responsible for creating/clearing then `TypeInfoStack`, I feel like it should be possible to store the stack within this serializer, and pass references to it around like the `TypeRegistry`, perhaps wrapping it all in a `ReflectContext`.

I don't see any problems with the `thread_local` implementation as-is beyond a general ick around global state; you're justification is perfectly valid. Just making a note here for a possible future PR.

Reply:
I don't fully remember unfortunately. 

I believe the reason was that the serializer would need to hold onto a `&mut TypeInfoStack` but we don't want to require the user to define this themselves.

We might be able to get around this by creating an `InternalReflectSerializer`. Then `ReflectSerializer` can create and own the stack, and then pass a mutable reference to it to the internal serializer.

That would be a breaking change though. I'll try to look into it and see if it will actually work.

Reply:
Just looked into this and it I believe it's possible to do this with an internal serializer, it wouldn't really work properly across threads.

This is because a `Serialize` impl only requires a `&self` reference. This means that two references to the same serializer could be sent to different threads and be serialized at the same time (maybe to send multiple network packets, save to separate files, etc.). Doing so would cause the stack to receive entries from both, throwing it completely off.

So for now, I think using a thread-local is probably the safest option.